lib/resty/redis.lua

```diff
@@ -351,14 +358,54 @@ local function _do_cmd(self, ...)
 end
 
 
-local function _check_subscribed(self, res)
+local function _check_unsubscribed(self, res)
     if type(res) == "table"
        and (res[1] == "unsubscribe" or res[1] == "punsubscribe")
-       and res[3] == 0
+    then
+        self._n_channel[res[1]] = self._n_channel[res[1]] - 1
+
+        local buffered_msg = rawget(self, "_buffered_msg")
+        if buffered_msg then
+            -- remove messages of unsubscribed channel
+            local msg_type =
+                (res[1] == "punsubscribe") and "pmessage" or "message"
+            local j = 1
+            for _, msg in ipairs(buffered_msg) do
+                if msg[1] == msg_type and msg[2] ~= res[2] then
+                    -- move messages to overwrite the removed ones
+                    buffered_msg[j] = msg
+                    j = j + 1
+                end
+            end
+
+            -- clear remain messages
+            for i = j, #buffered_msg do
+                buffered_msg[i] = nil
+            end
+
+            if #buffered_msg == 0 then
```

Comment:
```suggestion
            if j == 1 then
```

Reply:
@doujiang24 
The `buffered_msg` is cleared in the loop above while `j` doesn't change.